yarn-project/archiver/src/archiver/data_retrieval.ts

```diff
@@ -34,25 +34,29 @@ type DataRetrieval<T> = {
  * @param publicClient - The viem public client to use for transaction retrieval.
  * @param rollupAddress - The address of the rollup contract.
  * @param blockUntilSynced - If true, blocks until the archiver has fully synced.
- * @param currentL1BlockNum - Latest available block number in the ETH node.
  * @param searchStartBlock - The block number to use for starting the search.
+ * @param searchEndBlock - The highest block number that we should search up to.
  * @param expectedNextL2BlockNum - The next L2 block number that we expect to find.
  * @returns An array of L2 Blocks and the next eth block to search from
  */
 export async function retrieveBlocks(
   publicClient: PublicClient,
   rollupAddress: EthAddress,
   blockUntilSynced: boolean,
-  currentL1BlockNum: bigint,
   searchStartBlock: bigint,
+  searchEndBlock: bigint,
   expectedNextL2BlockNum: bigint,
 ): Promise<DataRetrieval<L2Block>> {
   const retrievedBlocks: L2Block[] = [];
   do {
-    if (searchStartBlock > currentL1BlockNum) {
+    if (searchStartBlock > searchEndBlock) {
       break;
     }
-    const l2BlockProcessedLogs = await getL2BlockProcessedLogs(publicClient, rollupAddress, searchStartBlock);
+    const unfilteredLogs = await getL2BlockProcessedLogs(publicClient, rollupAddress, searchStartBlock, searchEndBlock);
+    // ensure we only pulled logs from the expected block range
+    const l2BlockProcessedLogs = unfilteredLogs.filter(
+      log => log.blockNumber !== null && log.blockNumber >= searchStartBlock && log.blockNumber <= searchEndBlock,
+    );
```

Comment:
Seems like these checks should be redundant, but if we really don't trust the provider to return the correct block range, I'd move this check to within the `getL2BlockProcessedLogs` function, so we know that the getter already returns the correct data no matter where we call it from (same for other instances of this).

Reply:
I'll probably just get rid of the check. It was how I originally implemented the fix and then I saw that viem offered the `toBlock` parameter and used that.